src/common/src/array/compact_chunk.rs

```diff
@@ -0,0 +1,205 @@
+            debug_assert_eq!(prev.op(), Op::Delete);
+            debug_assert_eq!(self.latest.op(), Op::Insert);
+            (prev, &mut self.latest)
+        })
+    }
+}
+
+type OpRowMap<'a, 'b> =
+    HashMap<Prehashed<Project<'b, RowRef<'a>>>, OpRowMutRefTuple<'a>, BuildHasherDefault<Passthru>>;
+
+impl StreamChunkCompactor {
+    pub fn new(stream_key: Vec<usize>) -> Self {
+        Self {
+            stream_key,
+            chunks: vec![],
+        }
+    }
+
+    pub fn into_inner(self) -> (Vec<StreamChunk>, Vec<usize>) {
+        (self.chunks, self.stream_key)
+    }
+
+    pub fn push_chunk(&mut self, c: StreamChunk) {
+        self.chunks.push(c);
+    }
+
+    /// Compact a chunk by modifying the ops and the visibility of a stream chunk. All UPDATE INSERT
+    /// and UPDATE DELETE will be converted to INSERT and DELETE, and dropped according to
+    /// certain rules (see `merge_insert` and `merge_delete` for more details).
+
+    pub fn into_compacted_chunks(self) -> Vec<StreamChunk> {
+        let (chunks, key_indices) = self.into_inner();
+
+        let estimate_size = chunks.iter().map(|c| c.cardinality()).sum();
+        let mut chunks: Vec<(Vec<u64>, StreamChunkMut)> = chunks
+            .into_iter()
+            .map(|c| {
+                let hash_values = c
+                    .data_chunk()
+                    .get_hash_values(&key_indices, Crc32FastBuilder)
+                    .into_iter()
+                    .map(|hash| hash.value())
+                    .collect_vec();
+                (hash_values, StreamChunkMut::from(c))
+            })
+            .collect_vec();
+
+        let mut op_row_map: OpRowMap<'_, '_> = new_prehashed_map_with_capacity(estimate_size);
+        for (hash_values, c) in &mut chunks {
+            for (row, mut op_row) in c.to_rows_mut() {
+                if !op_row.vis() {
+                    continue;
+                }
+                op_row.set_op(op_row.op().normalize_update());
+                let hash = hash_values[row.index()];
+                let stream_key = row.project(&key_indices);
+                match op_row_map.entry(Prehashed::new(stream_key, hash)) {
+                    Entry::Vacant(v) => {
+                        v.insert(OpRowMutRefTuple {
+                            previous: None,
+                            latest: op_row,
+                        });
+                    }
+                    Entry::Occupied(mut o) => {
+                        if o.get_mut().push(op_row) {
+                            o.remove_entry();
+                        }
+                    }
+                }
+            }
+        }
+        for tuple in op_row_map.values_mut() {
+            if let Some((prev, latest)) = tuple.as_update_op() {
+                if prev.row_ref() == latest.row_ref() {
+                    prev.set_vis(false);
+                    latest.set_vis(false);
+                }
+            }
+        }
```

Comment:
I spent about 5 minutes to understand the logic here 🤣 It takes a very clever way to in-place update the visibility instead of rebuild new ones.

However, I think the motivation behind this design is because the compaction happens on `Project`, while due to my last comment, I oppose it. Instead, let's focus on the original user requirement and **do it on Sink fisrt**.

If so, considering the disordering problem of Sink, we can solve both problems with a single approach (as I have proposed in [Slack](https://risingwave-labs.slack.com/archives/C05QVSXDZPX/p1693972215979329))

- If sink_key != stream_key
  1. Shuffle the sink_key to stream_key
  2. Before or on SinkExecutor, do compaction for every sink_key
     - Collect all op & rows
     - Make them **offset** with each other
     - Output one of the 3 forms: `-`, `+`, `U- U+`

For example:

```
// sink_key = [a], stream_key = [a, b]
+(1,3)
+(1,2)
-(1,2)
-(1,1)
// Collect all op & rows
1 -> (+3, +2, -2, -1)
// Make them offset with each other 
1 -> (+3, , , -1)
// Output one of the 3 forms: `-`, `+`, `U- U+`
U-(1,3)
U+(1,1)
```

In this way, problem can be solved by building one hash table. Because this only happen on `Sink`, rebuilding Chunk is acceptable to me.

Besides, @st1page is considering that this clever implementation can be reused. But I didn't see any requirements for reusing it 🥹

Reply:
At least currently, we have two different use cases here
1. When sink's pk is different from the stream key
2. To compact the useless update value with the same values.

Reply:
> 2. To compact the useless update value with the same values.

Yeah, this makes sense, but 

- I do think it might be dangerous to reorder events in the process of streaming. After all, no operator does this now. But sorry I can't come up with a good example now.
- I hope to preserve the `U-/U+` rows. There are still lots of operators can benefit more or less from this. Reordering all deletes before inserts will completely break `U-/U+`.

--------

Alternatively, we can partially achieve this while preserving event orders - the solution is trivial: just hide the update with same before & after values. 

For example, set `vis = false` for these:

```
U- 1 2 'foo'
U+ 1 2 'foo'
```

but not these:

```
- 1 2 'foo'
... optionally, some other rows ...
+ 1 2 'foo'
```

Reply:
Yes, but this PR does not reorder any events :innocent:

Reply:
> Yes, but this PR does not reorder any events 😇

Of course. `StreamChunkCompactor` is not used at all. So how about removing this part until one day we need it?

Reply:
Yes, because it's now only used in `SinkExecutor`, not `Project`. I said so because you mentioned it can be reused (but not yet now).